### DISCUSSION TITLE
Simulate Trajectory following? [Closed]

### DISCUSSION THREAD

**Discussion by Doctor-N0** [do-mpc/do-mpc#269] [General]
Let's assume we have a very simple robot model:

<img src="https://latex.codecogs.com/svg.image?\begin{align*}\frac{\partial&space;x}{\partial&space;t}&space;&=&space;v&space;\cdot&space;\cos(\theta)\\\frac{\partial&space;y}{\partial&space;t}&space;&=&space;v&space;\cdot&space;\sin(\theta)\\\frac{\partial&space;\theta}{\partial&space;t}&space;&=&space;\omega\end{align*}" title="https://latex.codecogs.com/svg.image?\begin{align*}\frac{\partial x}{\partial t} &= v \cdot \cos(\theta)\\\frac{\partial y}{\partial t} &= v \cdot \sin(\theta)\\\frac{\partial \theta}{\partial t} &= \omega\end{align*}" />

With the states

<img src="https://latex.codecogs.com/svg.image?\begin{align*}x&space;=&space;[x,y,\theta]^{T}\end{align*}" title="https://latex.codecogs.com/svg.image?\begin{align*}x = [x,y,\theta]^{T}\end{align*}" />

And Input
<img src="https://latex.codecogs.com/svg.image?u&space;=&space;[v,\omega]^{T}" title="https://latex.codecogs.com/svg.image?u = [v,\omega]^{T}" />

`x`  and `y`  are the coordinates of the robot in a x-y-Plane, `theta`  is its rotation. The input  `v`  is the velocity is driving direction and  `omega` is the heading angle.

Now I would like my robot to follow a trajectory, e.g. a straight line
<img src="https://latex.codecogs.com/svg.image?x_{ref}&space;=&space;t,&space;\:&space;y_{ref}=2t" title="https://latex.codecogs.com/svg.image?x_{ref} = t, \: y_{ref}=2t" />
or maybe a Lissajous curve
<img src="https://latex.codecogs.com/svg.image?x_{ref}&space;=&space;sin(t),&space;\:&space;y_{ref}=sin(2t)" title="https://latex.codecogs.com/svg.image?x_{ref} = sin(t), \: y_{ref}=sin(2t)" />

It this possible using do-mpc?

Thank you for your answers!

PS: Absolutely amazed by do-mpc so far. Great project!

**Comment by 4flixt** [comment]
Hi,

first: I have converted that to a discussion (instead of an issue). Regarding your question: Absolutely, that is exactly what do-mpc is for. Why don't you try this out?

Best,
4flixt

**Reply by Doctor-N0** [reply]
Could you please give me a hint where I might start looking? I assume that I have to assign the control input `u` (trajectory) to the simulator, but in the examples, I just found a step input via `mpc.make_step(x0)`.

**Reply by 4flixt** [reply]
Dear @Doctor-N0, can you have a look at the examples on our homepage, e.g. https://www.do-mpc.com/en/latest/getting_started.html
I think that should make it easy to adapt to your case.

Best,
4flixt

**Reply by Doctor-N0** [reply]
Thanks, I'll do that!

**Reply by taufeey** [reply]
![CSTR Output](https://user-images.githubusercontent.com/45310462/219856553-852cc3a0-72b5-4fd0-b052-19d48952c280.png)
@4flixt @Doctor-N0 

Dears,
Please find the attachment.
I tried to implement CSTR in the guided way static graph i was able to obtained on jupyter Notebook. But unfortunately still lacking to get runtime graph for prediction and set point. As mentioned and displayed on the do-mpc example page.

help needed for the proper implementation of the below mentioned part: 

Some “cosmetic” modifications are easily achieved with the structure pred_lines and result_lines.

# Update properties for all prediction lines:
for line_i in mpc_graphics.pred_lines.full:
    line_i.set_linewidth(2)
# Highlight nominal case:
for line_i in np.sum(mpc_graphics.pred_lines['_x', :, :,0]):
    line_i.set_linewidth(5)
for line_i in np.sum(mpc_graphics.pred_lines['_u', :, :,0]):
    line_i.set_linewidth(5)
for line_i in np.sum(mpc_graphics.pred_lines['_aux', :, :,0]):
    line_i.set_linewidth(5)

# Add labels
label_lines = mpc_graphics.result_lines['_x', 'C_a']+mpc_graphics.result_lines['_x', 'C_b']
ax[0].legend(label_lines, ['C_a', 'C_b'])
label_lines = mpc_graphics.result_lines['_x', 'T_R']+mpc_graphics.result_lines['_x', 'T_K']
ax[1].legend(label_lines, ['T_R', 'T_K'])

fig.align_ylabels()
After importing the necessary package:

from matplotlib.animation import FuncAnimation, ImageMagickWriter
We obtain the animation with:

def update(t_ind):
    print('Writing frame: {}.'.format(t_ind), end='\r')
    mpc_graphics.plot_results(t_ind=t_ind)
    mpc_graphics.plot_predictions(t_ind=t_ind)
    mpc_graphics.reset_axes()
    lines = mpc_graphics.result_lines.full
    return lines

n_steps = mpc.data['_time'].shape[0]


anim = FuncAnimation(fig, update, frames=n_steps, blit=True)

gif_writer = ImageMagickWriter(fps=5)
anim.save('anim_CSTR.gif', writer=gif_writer)


furthermore if i try to implement same code in pycharm it executed the iterations properly but graph is not appearing.

Best regards & thanks

**Comment by Doctor-N0** [comment]
The answer was actually quite simple, using `t_now` in the `tvp_fun_mpc` function. Here's a Minimal Working Example of a Straight Line or a Lissajous Curve:

```
tvp_struct_mpc = mpc.get_tvp_template()

def tvp_fun_mpc(t_now):
    # straight line
    tvp_struct_mpc['_tvp',:,'x_set_point'] =  0.5*t_now
    tvp_struct_mpc['_tvp',:,'y_set_point'] =      t_now
    
    # Lissajous Curve
    # tvp_struct_mpc['_tvp',:,'x_set_point'] = 2 * sin(0.5*t_now)
    # tvp_struct_mpc['_tvp',:,'y_set_point'] = 2 * sin(1*t_now)
    return tvp_struct_mpc

# Pass tvp function to mpc instance
mpc.set_tvp_fun(tvp_fun_mpc)
```
